### PR TITLE
Improved LIS and added Rabbit solution

### PR DESCRIPTION
**Pull request by SansPapyrus683**
Added some examples, elaborated on the proof, and added Python support.

### CODE REVIEW

**Comment by dolphingarlic** [comment]
The LIS is normally strictly increasing though

File: content/4_Gold/LIS.mdx [hunk 1]
```diff
@@ -144,20 +155,38 @@ int find_lis(vector<int> a) {
...
+		}
 	}
 	return dp.size();
 }
 ```
 
 </JavaSection>
 
+<PySection>
+# list from typing module
```
Comment (dolphingarlic) [inline content/4_Gold/LIS.mdx:178]: Wrap the code in triple back ticks

File: content/4_Gold/LIS.mdx [hunk 2]
```diff
@@ -144,20 +155,38 @@ int find_lis(vector<int> a) {
...
 	ArrayList<Integer> dp = new ArrayList<Integer>();
 	for (int i : a) {
 		int pos = Collections.binarySearch(dp, i);
-		if(pos < 0) pos = Math.abs(pos + 1);
-		if (pos == dp.size()) dp.add(i);
-		else dp.set(pos, i);
+		if (pos < 0) {
+			pos = Math.abs(pos + 1);
+		}
+		if (pos == dp.size()) {  // we can have a new, longer IS!
```
Comment (dolphingarlic) [inline content/4_Gold/LIS.mdx:165]: Rather say "increasing subsequence" instead of "IS"

File: content/4_Gold/LIS.mdx [hunk 3]
```diff
@@ -132,8 +140,11 @@ int find_lis(vector<int> a) {
 	vector<int> dp;
 	for (int i : a) {
 		int pos = lower_bound(dp.begin(), dp.end(), i) - dp.begin();
-		if (pos == dp.size()) dp.push_back(i);
-		else dp[pos] = i;
+		if (pos == dp.size()) {  // we can have a new, longer IS!
+			dp.push_back(i);
+		} else {  // oh ok, at least we can make the ending element smaller
+			dp[pos] = i;
+		}
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:147]: Fine as is, no?

File: content/4_Gold/LIS.mdx [hunk 4]
```diff
@@ -144,20 +155,38 @@ int find_lis(vector<int> a) {
...
-		if (pos == dp.size()) dp.add(i);
-		else dp.set(pos, i);
+		if (pos < 0) {
+			pos = Math.abs(pos + 1);
+		}
+		if (pos == dp.size()) {  // we can have a new, longer IS!
+			dp.add(i);
+		} else {  // oh ok, at least we can make the ending element smaller
+			dp.set(pos, i);
+		}
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:169]: Seems unnecessarily long compared to the 3 concise lines D:

**Review by dolphingarlic — CHANGES_REQUESTED** [review]

File: content/4_Gold/LIS.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx:178]: Oh yeah, forgot about that.

File: content/4_Gold/LIS.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx:147]: I feel like just adding braces make it look better. I can still remove them if you want.

Comment (frodakcin) [reply content/4_Gold/LIS.mdx]: Oh, I personally think braces for 1-liners looks messy. Generally I leave stuff like this to the author of the code though

File: content/4_Gold/LIS.mdx [hunk 5]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
 elements of $A$ with an increasing sequence of length $j$ ending on it (or
 $\infty$ if there is no such element).
 
-**Lemma 1:** $L_i$ forms a non-decreasing sequence.
+For example, let our array be `2 18 7 20 18 5 18 15 13 19 9`.
+
+In this case, $L_8$ would be `[2, 5, 15]`, some example sequences satisfying
```
Comment (dolphingarlic) [inline content/4_Gold/LIS.mdx:102]:
```suggestion
In this case, $L_8$ would be `[2, 5, 15]`. Some example sequences satisfying
```

File: content/4_Gold/LIS.mdx [hunk 6]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
+these being `[2]`, `[2, 5]`, and `[2, 5, 15]`.
```
Comment (dolphingarlic) [inline content/4_Gold/LIS.mdx:103]:
```suggestion
these are `[2]`, `[2, 5]`, and `[2, 5, 15]`.
```

File: content/4_Gold/LIS.mdx [hunk 7]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
+
+**Lemma 1:** $L_i$ forms a strictly increasing sequence.
 
 **Proof:** Assume for a contradiction that for some $j$, we have
-$L_i[j] > L_i[j + 1]$. However, this is impossible because an increasing
-sequence of length $j + 1$ ending on some element implies that there is also an
-increasing sequence of length $j$ ending on that same sequence.
+$L_i[j] \geq L_i[j + 1]$. Define $a_i$ ($0 \leq i \leq j$) to be the $i$th
```
Comment (dolphingarlic) [inline content/4_Gold/LIS.mdx:108]:
```suggestion
$L_i[j] \geq L_i[j + 1]$. Define $a_i$ ($0 \leq i \leq j$) to be the $i$-th
```

File: content/4_Gold/LIS.mdx [hunk 8]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
+element of any increasing subsequence that ends with $L_i[j + 1]$. We can have
+an increasing subsequence on length $j$ that ends on $a_{j - 1}$ which is less
```
Comment (dolphingarlic) [inline content/4_Gold/LIS.mdx:110]:
```suggestion
an increasing subsequence of length $j$ that ends on $a_{j - 1}$, which is less
```

File: content/4_Gold/LIS.mdx [hunk 9]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
-index $j$ such that $L_i[j] \geq A[i + 1]$.
+index $j$ (with 0-indexing) such that $L_i[j] \geq A[i + 1]$.
 
-**Proof:** Firstly, since $A[i + 1] > L_i[j - 1]$, there is an increasing
-sequence of length $j$ ending on $A[i + 1]$. By Lemma 1, $L_i$ is
-non-decreasing, so $L_i[k] \geq A[i + 1]$ for all $k \geq j$. This means that
-the length of the LIS is $j$.
+**Proof:** Let $j$ be defined as it is in the statement. First of all, since
+$A[i + 1] > L_i[j - 1]$, there is an increasing sequence of length $j$ ending on
+$A[i + 1]$ as we can append $A[i + 1]$ to the end of the sequence of length
```
Comment (dolphingarlic) [inline content/4_Gold/LIS.mdx:118]:
```suggestion
$A[i + 1]$, as we can append $A[i + 1]$ to the end of the sequence of length
```

File: content/4_Gold/LIS.mdx [hunk 10]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
+$j - 1. By Lemma 1, $L_i$ is strictly increasing, so $L_i[k] \geq A[i + 1]$ for
```
Comment (dolphingarlic) [inline content/4_Gold/LIS.mdx:119]:
```suggestion
$j - 1$. By Lemma 1, $L_i$ is strictly increasing, so $L_i[k] \geq A[i + 1]$ for
```

**Review by dolphingarlic — CHANGES_REQUESTED** [review]

File: content/4_Gold/LIS.mdx (hunk 3 above)
Comment (mizlan) [reply content/4_Gold/LIS.mdx]: My two cents: leave the braces in so control flow is easier to understand.

File: content/4_Gold/LIS.mdx (hunk 4 above)
Comment (mizlan) [reply content/4_Gold/LIS.mdx]: They're equally concise, though, because the code is the same. One just makes control flow more obvious.

Comment (dolphingarlic) [reply content/4_Gold/LIS.mdx]: It's a single statement though. I'd argue that making the code more vertically concise is better for readability because it's more compact and elegant

Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx]: I mean with monitors as big as they are, being vertically concise shouldn't be that big of a concern (also I compacted the `Math.abs` part).

File: content/4_Gold/LIS.mdx [hunk 11]
```diff
@@ -144,20 +154,40 @@ int find_lis(vector<int> a) {
...
 	return dp.size();
 }
 ```
 
 </JavaSection>
 
+<PySection>
+
+```py
+# list from typing module
```
Comment (dolphingarlic) [inline content/4_Gold/LIS.mdx:177]: Do you think it's better to explicitly import these modules?

Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx:177]: C++ and Java didn't, so I'm not sure.

Comment (dolphingarlic) [reply content/4_Gold/LIS.mdx:177]: Ok then it's fine

File: content/4_Gold/LIS.mdx (hunk 4 above)
Comment (mizlan) [reply content/4_Gold/LIS.mdx]: > more compact and elegant

No, the former code is not, in fact, more elegant.

Comment (dolphingarlic) [reply content/4_Gold/LIS.mdx]: How so?

Comment (frodakcin) [reply content/4_Gold/LIS.mdx]: > control flow more obvious

Yes, in the original way you have a series of (condition -> statement). This exactly contributes to why concise is cleaner.

The alternative requires you to connect condition & statement between multiple lines. Of course this would be preferred for a set of substantial, multiline statement, but not when it is a sleek one-liner.

Comment (frodakcin) [reply content/4_Gold/LIS.mdx]: In the end settling with the choice of the original author is just more respectful, especially when there is no significant reason to change.

Comment (mizlan) [reply content/4_Gold/LIS.mdx]: By virtue of the semantics being the exact same. The indentation present in the latter, however, makes the control flow easier to digest—that is the point of indentation. Braces are a good practice anyway and doesn't make it any less "elegant" or readable.

File: content/4_Gold/LIS.mdx [hunk 12]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
 
 **Lemma 2:** The length of the LIS ending on $A[i + 1]$ is equal to the least
-index $j$ such that $L_i[j] \geq A[i + 1]$.
+index $j$ (with 0-indexing) such that $L_i[j] \geq A[i + 1]$.
 
-**Proof:** Firstly, since $A[i + 1] > L_i[j - 1]$, there is an increasing
-sequence of length $j$ ending on $A[i + 1]$. By Lemma 1, $L_i$ is
-non-decreasing, so $L_i[k] \geq A[i + 1]$ for all $k \geq j$. This means that
-the length of the LIS is $j$.
+**Proof:** Let $j$ be defined as it is in the statement. First of all, since
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:116]:
```suggestion
**Proof:** Since
```

File: content/4_Gold/LIS.mdx [hunk 13]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
+$A[i + 1] > L_i[j - 1]$, there is an increasing sequence of length $j$ ending on
+$A[i + 1]$, as we can append $A[i + 1]$ to the end of the sequence of length
+$j - 1$. By Lemma 1, $L_i$ is strictly increasing, so $L_i[k] \geq A[i + 1]$ for
+all $k \geq j$. This means that the length of the LIS is $j$.
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:120]:
```suggestion
all $k \geq j$. This means that there cannot be a LIS ending at $A[i+1]$ longer than $j$.
```

File: content/4_Gold/LIS.mdx [hunk 14]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
 
-**Lemma 3:** Exactly 1 element differs between $L_i$ and $L_{i + 1}$.
+**Lemma 3:** At most 1 element differs between $L_i$ and $L_{i + 1}$.
 
-**Proof:** Obviously, we need to set $L_{i + 1}[j]$ to be $A[i + 1]$ since
-$L_i[j] \geq A[i + 1]$. We don't update anything else though, since
-$A[i + 1] > L_i[k]$ for all $k < j$ and there are no increasing sequences ending
-on $A[i + 1]$ of length greater than $j$.
+**Proof:** Let $j$ be defined as it was in the previous lemma. We need to set
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:124]:
```suggestion
**Proof:** Let $j$ be defined as in Lemma 2. We need to set
```

File: content/4_Gold/LIS.mdx [hunk 15]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
+$L_{i + 1}[j]$ to be $A[i + 1]$ since $A[i + 1] \leq L_i[j]$. We don't update
+anything else though, since $A[i + 1] > L_i[k]$ for all $k < j$ and there are no
+increasing sequences ending on $A[i + 1]$ of length greater than $j$.
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:127]:
```suggestion
$L_{i + 1}[j]$ to be $A[i + 1]$ since $A[i + 1] \leq L_i[j]$. $L_{i + 1}[k] = L_i[k]$
for all $k \neq j$, though, since $A[i + 1] > L_i[k]$ for all $k < j$ and there are no
increasing sequences ending on $A[i + 1]$ for $k > j$.
```

Maybe this helps?

File: content/4_Gold/LIS.mdx [hunk 16]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
 
 To find and update the described $j$ in $\mathcal{O}(\log N)$ time, we can use a
-`std::vector` and `std::lower_bound`. Alternatively, we can use a `std::set`
-(demonstrated in the solution for PCB).
+list with binary search, or we can use a sorted set (as demonstrated in the
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:130]: I think "ordered set" is more accurate

**Review by frodakcin — CHANGES_REQUESTED** [review]
Wording changes are kinda sketchy -- I'm sure there's better ways to phrase stuff but you can pick & choose the ones ive suggested

File: content/4_Gold/LIS.mdx [hunk 17]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
 elements of $A$ with an increasing sequence of length $j$ ending on it (or
 $\infty$ if there is no such element).
 
-**Lemma 1:** $L_i$ forms a non-decreasing sequence.
+For example, let our array be `2 18 7 20 18 5 18 15 13 19 9`.
+
+In this case, $L_8$ would be `[2, 5, 15]`. Some example sequences satisfying
```
[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx:102]: Actually yeah, L_8[j] is the minimal ending element of a sequence that has length j + 1 instead of length j.

File: content/4_Gold/LIS.mdx (hunk 12 above)
Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx:116]: You sure? I feel like that makes it a bit more confusing.

Comment (frodakcin) [reply content/4_Gold/LIS.mdx:116]: It's fine as is, but seems redundant

File: content/4_Gold/LIS.mdx [hunk 18]
```diff
@@ -97,31 +97,38 @@ Let $L_i$ be an array where $L_i[j]$ is the smallest element from the first $i$
...
+**Lemma 1:** $L_i$ forms a strictly increasing sequence.
 
 **Proof:** Assume for a contradiction that for some $j$, we have
-$L_i[j] > L_i[j + 1]$. However, this is impossible because an increasing
-sequence of length $j + 1$ ending on some element implies that there is also an
-increasing sequence of length $j$ ending on that same sequence.
+$L_i[j] \geq L_i[j + 1]$. Define $a_i$ ($0 \leq i \leq j$) to be the $i$-th
+element of any increasing subsequence that ends with $L_i[j + 1]$. We can have
+an increasing subsequence of length $j$ that ends on $a_{j - 1}$, which is less
+than $a_j = L_i[j + 1]$, so there is a contradiction.
```
[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx:111]: >  (Note, $L_i[j]$ may be less than $a_{j-1}$ for all sequences $a$ of length $j+1$.)

Wait, what does this mean?

[the remark this answers is not in the document]
Comment (frodakcin) [reply content/4_Gold/LIS.mdx:111]: Probably unnecessary, but it basically means: the optimal answer for a subsequence of length j isn't necessary the first $j$ terms of an optimal subsequence of length j+1

**Comment by frodakcin** [comment]
Uhh, not sure TWE you changed stuff, but I think it's just easier to say that L_i[0] is infinite or that you wrote it out 1-indexed

**Comment by SansPapyrus683** [comment]
TWE? Also I think it's easier to just keep that definition, since it connects directly with the code.

**Comment by frodakcin** [comment]
Oh, sure

File: content/4_Gold/LIS.mdx [hunk 19]
```diff
@@ -105,10 +105,12 @@ these are $[2]$, $[2, 5]$, and $[2, 5, 15]$.
...
-$L_i[j] \geq L_i[j + 1]$. Define $a_i$ ($0 \leq i \leq j$) to be the $i$-th
-element of any increasing subsequence that ends with $L_i[j + 1]$. We can have
-an increasing subsequence of length $j$ that ends on $a_{j - 1}$, which is less
-than $a_j = L_i[j + 1]$, so there is a contradiction.
+$L_i[j] \geq L_i[j + 1]$. Let $a_0, a_1, \dots, a_j$ be any increasing subsequence
+of length $j+1$ ending on $L_i[j + 1]$ (so $a_j=L_i[j + 1]$). Notice that
+$a_0, a_1, \dots, a_{j-1}$ is an increasing subsequence of length $j$.
+Thus, $L_i[j]$ cannot be greater than $a_{j-1}$.
+From that, we can write this equation: $L_i[j] \leq a_{j-1} < a_j = L_i[j+1]$,
+which is exactly what we wanted.
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx]: Uhh, is this still accurate per the new definition of L_i[j]?

I believe indicies should be shifted by 1

Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx]: Oh yeah, I'll change that.

File: content/4_Gold/LIS.mdx [hunk 20]
```diff
@@ -93,35 +93,46 @@ We can do much better than this though!
...
-**Proof:** Firstly, since $A[i + 1] > L_i[j - 1]$, there is an increasing
-sequence of length $j$ ending on $A[i + 1]$. By Lemma 1, $L_i$ is
-non-decreasing, so $L_i[k] \geq A[i + 1]$ for all $k \geq j$. This means that
-the length of the LIS is $j$.
+**Proof:** Let $j$ be defined as it is in the statement. First of all, since
+$A[i + 1] > L_i[j - 1]$, there is an increasing sequence of length $j$ ending on
+$A[i + 1]$, as we can append $A[i + 1]$ to the end of the sequence of length
+$j - 1$. By Lemma 1, $L_i$ is strictly increasing, so $L_i[k] \geq A[i + 1]$ for
+all $k \geq j$. This means that there cannot be a LIS ending at $A[i + 1]$
+longer than $j$.
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:124]: Shift indicies

File: content/4_Gold/LIS.mdx [hunk 21]
```diff
@@ -93,35 +93,46 @@ We can do much better than this though!
...
 
-**Proof:** Obviously, we need to set $L_{i + 1}[j]$ to be $A[i + 1]$ since
-$L_i[j] \geq A[i + 1]$. We don't update anything else though, since
-$A[i + 1] > L_i[k]$ for all $k < j$ and there are no increasing sequences ending
-on $A[i + 1]$ of length greater than $j$.
+**Proof:** Let $j$ be defined as it was in Lemma 2. We need to set
+$L_{i + 1}[j]$ to be $A[i + 1]$ since $A[i + 1] \leq L_i[j]$.
+$L_{i + 1}[k] = L_i[k]$ for all $k \neq j$, though, since $A[i + 1] > L_i[k]$
+for all $k < j$ and there are no increasing sequences ending on $A[i + 1]$ for
+$k > j$.
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:131]: Does this need index shifting too?

Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx:131]: Don't think so...

File: content/4_Gold/LIS.mdx [hunk 22]
```diff
@@ -132,8 +143,11 @@ int find_lis(vector<int> a) {
 	vector<int> dp;
 	for (int i : a) {
 		int pos = lower_bound(dp.begin(), dp.end(), i) - dp.begin();
-		if (pos == dp.size()) dp.push_back(i);
-		else dp[pos] = i;
+		if (pos == dp.size()) {  // we can have a new, longer increasing subsequence!
+			dp.push_back(i);
+		} else {  // oh ok, at least we can make the ending element smaller
+			dp[pos] = i;
+		}
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:150]: Leave this as is.

File: content/4_Gold/LIS.mdx [hunk 23]
```diff
@@ -144,20 +158,40 @@ int find_lis(vector<int> a) {
...
 		int pos = Collections.binarySearch(dp, i);
-		if(pos < 0) pos = Math.abs(pos + 1);
-		if (pos == dp.size()) dp.add(i);
-		else dp.set(pos, i);
+		pos = pos < 0 ? Math.abs(pos + 1) : pos;
+		if (pos == dp.size()) {  // we can have a new, longer increasing subsequence!
+			dp.add(i);
+		} else {  // oh ok, at least we can make the ending element smaller
+			dp.set(pos, i);
+		}
```
Comment (frodakcin) [inline content/4_Gold/LIS.mdx:170]: Leave as is

(I don't believe there is a compelling reason to change to camelcase, right?)

**Review by frodakcin — CHANGES_REQUESTED** [review]
TBH I briefly skimmed through the explanation -- in all honesty the proofs for the lemmas only have to be accurate but not necessarily as clear/understandable as would be ideal. The current algo is quite complex and probably needs an image/animation to make clear. I don't think wording changes will fix it.

tl;dr: revert cpp/java code; explanation is probably good enough

File: content/4_Gold/LIS.mdx (hunk 23 above)
Comment (SansPapyrus683) [reply content/4_Gold/LIS.mdx:170]: The convention for Java is camelCase. It's always good to be consistent.

**Comment by SansPapyrus683** [comment]
Oops- For some random reason I thought this was merged already. I hope it's ok to have PR with 2 changes.
